error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ComparatorRules.java

```diff
@@ -314,14 +347,46 @@ T after(T value1, T value2, Comparator<? super T> cmp) {
    * Avoid unnecessary creation of a {@link Stream} to determine the maximum of a known collection
    * of values.
    */
-  static final class MaxOfVarargs<T> {
+  static final class MaxOfArray<S, T extends S> {
+    @BeforeTemplate
+    T before(T[] array, Comparator<S> cmp) {
+      return Arrays.stream(array).max(cmp).orElseThrow();
+    }
+
+    @AfterTemplate
+    T after(T[] array, Comparator<S> cmp) {
+      return Collections.max(Arrays.asList(array), cmp);
+    }
+  }
+
+  /**
+   * Avoid unnecessary creation of a {@link Stream} to determine the maximum of a known collection
+   * of values.
+   */
+  static final class MaxOfCollection<S, T extends S> {
```

Comment:
Looking at the unit test and usage of `Comparator#naturalOrder`, it made me wonder if we can re-write the case when the natural order is used as a comparator to the `Collections#max(collection)`.
```patch
===================================================================
diff --git a/error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ComparatorRulesTestOutput.java b/error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ComparatorRulesTestOutput.java
--- a/error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ComparatorRulesTestOutput.java	(revision 49825759b2310bc0ea8fd72d8e4842f9c5460313)
+++ b/error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ComparatorRulesTestOutput.java	(date 1723313983688)
@@ -138,8 +138,12 @@
     return Collections.max(Arrays.asList(new String[0]), naturalOrder());
   }
 
-  String testMaxOfCollection() {
-    return Collections.max(ImmutableSet.of("foo", "bar"), naturalOrder());
+  String testMaxOfCollectionComparator() {
+    return Collections.max(ImmutableSet.of("foo", "bar"), reverseOrder());
+  }
+
+  String testMaxOfCollectionNaturalOrder() {
+    return Collections.max(ImmutableSet.of("foo", "bar"));
   }
 
   int testMaxOfVarargs() {
Index: error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ComparatorRulesTestInput.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>ISO-8859-1
===================================================================
diff --git a/error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ComparatorRulesTestInput.java b/error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ComparatorRulesTestInput.java
--- a/error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ComparatorRulesTestInput.java	(revision 49825759b2310bc0ea8fd72d8e4842f9c5460313)
+++ b/error-prone-contrib/src/test/resources/tech/picnic/errorprone/refasterrules/ComparatorRulesTestInput.java	(date 1723313983685)
@@ -147,7 +147,11 @@
     return Arrays.stream(new String[0]).max(naturalOrder()).orElseThrow();
   }
 
-  String testMaxOfCollection() {
+  String testMaxOfCollectionComparator() {
+    return ImmutableSet.of("foo", "bar").stream().max(reverseOrder()).orElseThrow();
+  }
+
+  String testMaxOfCollectionNaturalOrder() {
     return ImmutableSet.of("foo", "bar").stream().max(naturalOrder()).orElseThrow();
   }
 
Index: error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ComparatorRules.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ComparatorRules.java b/error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ComparatorRules.java
--- a/error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ComparatorRules.java	(revision 49825759b2310bc0ea8fd72d8e4842f9c5460313)
+++ b/error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ComparatorRules.java	(date 1723313988286)
@@ -363,7 +363,7 @@
    * Avoid unnecessary creation of a {@link Stream} to determine the maximum of a known collection
    * of values.
    */
-  static final class MaxOfCollection<S, T extends S> {
+  static final class MaxOfCollectionComparator<S, T extends S> {
     @BeforeTemplate
     T before(Collection<T> collection, Comparator<S> cmp) {
       return collection.stream().max(cmp).orElseThrow();
@@ -375,6 +375,22 @@
     }
   }
 
+  /**
+   * Avoid unnecessary creation of a {@link Stream} to determine the maximum of a known collection
+   * of values.
+   */
+  static final class MaxOfCollectionNaturalOrder<T extends Comparable<? super T>> {
+    @BeforeTemplate
+    T before(Collection<T> collection) {
+      return collection.stream().max(naturalOrder()).orElseThrow();
+    }
+
+    @AfterTemplate
+    T after(Collection<T> collection) {
+      return Collections.max(collection);
+    }
+  }
+
   /**
    * Avoid unnecessary creation of a {@link Stream} to determine the maximum of a known collection
    * of values.
```

Reply:
Or perhaps 
```
Collections.max(collection, naturalOrder());
```
to 
```
Collections.max(collection);
```
could be its own rule 👀

Reply:
Indeed, it should be a separate rule, for maximum applicability. And likewise for `min`, of course, and with `naturalOrder()` and `reverseOrder()` variants. Are you up for a PR? :smile:

Reply:
Makes sense 👌 

> Are you up for a PR? 😄

Always 🫡

Reply:
https://github.com/PicnicSupermarket/error-prone-support/pull/1297